src/middlewared/middlewared/main.py

```diff
@@ -392,6 +393,145 @@ def read_write():
+            the shell pty file descriptor.
+            """
+            while True:
+                try:
+                    get = self.input_queue.get(timeout=1)
+                    os.write(master_fd, get)
+                except queue.Empty:
+                    # If we timeout waiting in input query lets make sure
+                    # the shell process is still alive
+                    try:
+                        os.kill(self.shell_pid, 0)
+                    except ProcessLookupError:
+                        break
+
+        t_reader = threading.Thread(target=reader, daemon=True)
+        t_reader.start()
+
+        t_writer = threading.Thread(target=writer, daemon=True)
+        t_writer.start()
+
+        # Wait for shell to exit
+        os.waitpid(self.shell_pid, 0)
+
+        t_reader.join()
+        t_writer.join()
+
+
+class ShellApplication(object):
+
+    def __init__(self, middleware):
+        self.middleware = middleware
+
+    async def ws_handler(self, request):
+        ws = web.WebSocketResponse()
+        await ws.prepare(request)
+
+        # Each connection will have its own input queue
+        input_queue = queue.Queue()
+        t_worker = None
+        authenticated = False
+
+        async for msg in ws:
+            if authenticated:
+                # Add content of every message received in input queue
+                try:
+                    input_queue.put(msg.data.encode())
+                except UnicodeEncodeError:
+                    # Should we handle Encode error?
+                    # xterm.js seems to operate with the websocket in text mode,
+                    pass
+            else:
+                try:
+                    data = json.loads(msg.data)
+                except json.decoder.JSONDecodeError:
+                    continue
+
+                token = data.get('token')
+                if not token:
+                    continue
+
+                token = await self.middleware.call('auth.get_token', token)
+                if not token:
+                    ws.send_json({
+                        'msg': 'failed',
+                        'error': {
+                            'error': errno.EACCES,
+                            'reason': 'Invalid token',
+                        }
+                    })
+                    continue
+
+                authenticated = True
+                ws.send_json({
+                    'msg': 'connected',
+                })
+                t_worker = ShellWorkerThread(ws=ws, input_queue=input_queue)
+                t_worker.start()
+
+        # If connection was not authenticated, return earlier
+        if not authenticated:
+            return ws
+
+        # If connection has been closed lets make sure shell is killed
+        if t_worker.shell_pid:
+            try:
+                os.kill(t_worker.shell_pid, signal.SIGTERM)
+            except ProcessLookupError:
+                pass
+
```

Comment:
At that point the connection from client will be closed regardless and the coroutine/thread may hang indefinitely. However I am not sure how nor if its worth handling that case?

Do you have a proposed solution? I guess we could try wait a few and try to SIGKILL it and eventually release the thread if nothing works. I am not a fan of time.sleep so we would probably need to setup kqueue on pid, then things start to escalate =p